### PR TITLE
Change the `||` operator and the `array_cat` scalar to return NULL when all args are NULL

### PR DESCRIPTION
**Pull request by seut**
See commits.

### CODE REVIEW

File: libs/sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java [hunk 1]
```diff
@@ -1991,10 +1991,16 @@ public Node visitBitwiseBinary(SqlBaseParser.BitwiseBinaryContext context) {
 
     @Override
     public Node visitConcatenation(SqlBaseParser.ConcatenationContext context) {
+        var left = (Expression) visit(context.left);
+        var right = (Expression) visit(context.right);
+        if (left instanceof NullLiteral && right instanceof NullLiteral) {
+            return NullLiteral.INSTANCE;
+        }
```
Comment (seut) [inline libs/sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java:1998]: The main problem here is that the `null || null -> null` operator behaves different that `concat(null, null)  -> empty string`. The only other solution I'd come up with is passing the information that this was original an operator somehow to the function implementation which sounds complex.

File: libs/sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java [hunk 2]
```diff
@@ -1991,10 +1991,16 @@ public Node visitBitwiseBinary(SqlBaseParser.BitwiseBinaryContext context) {
...
+
         return new FunctionCall(
-            QualifiedName.of("concat"), List.of(
-            (Expression) visit(context.left),
-            (Expression) visit(context.right)));
+            QualifiedName.of("concat"),
+            List.of(left, right)
+        );
```
Comment (mfussenegger) [inline libs/sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java:2003]: This won't work if accessing columns. E.g. in PostgreSQL:

```
select x || y from unnest(array[null, 'a', null, 'c'], array[null, null, 'b', 'd']) as t (x, y);
 ?column?
----------
 💩
 💩
 💩
 cd
(4 rows)
```


With this change in CrateDB:

```
cr> select x || y from unnest(array[null, 'a', null, 'c'], array[null, null, 'b', 'd']) as t (x, y);
+--------------+
| concat(x, y) |
+--------------+
|              |
| a            |
| b            |
| cd           |
+--------------+
SELECT 4 rows in set (0.009 sec)
```

I'd revert this part, as it is not worth optimizing the uncommon case (`null || null`) and it actually slows down the more common case (working with columns).

We probably need to introduce a `_||` or `_concat` scalar that's similar to the `ConcatFunction` except with changed null handling.

File: server/src/main/java/io/crate/expression/scalar/ArrayCatFunction.java [hunk 3]
```diff
@@ -38,12 +39,28 @@
...
+import io.crate.types.UndefinedType;
 
 public class ArrayCatFunction extends Scalar<List<Object>, List<Object>> {
 
     public static final String NAME = "array_cat";
 
     public static void register(Functions.Builder module) {
+        // We need to register a concrete function implementation for the `NULL` arguments.
+        // Otherwise, the arguments type will cast to and `ARRAY(UNDEFINED)` which causes a validation error as
+        // the inner type of the array arguments cannot both be `UNDEFINED`.
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/scalar/ArrayCatFunction.java:51]: Would there be any downsides to allowing `array(undefined)` instead? I could imagine this case occuring more than just once, and it would avoid having the extra registration (plus the confusing error message containing `Possible candidates: array_cat(undefined, undefined):undefined`)

File: server/src/main/java/io/crate/expression/scalar/ArrayCatFunction.java [hunk 4]
```diff
@@ -38,12 +39,28 @@
...
+        module.add(
+            Signature.builder(NAME, FunctionType.SCALAR)
+                .argumentTypes(
+                    UndefinedType.INSTANCE.getTypeSignature(),
+                    UndefinedType.INSTANCE.getTypeSignature())
+                .returnType(UndefinedType.INSTANCE.getTypeSignature())
+                .features(Feature.DETERMINISTIC, Feature.NULLABLE)
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/scalar/ArrayCatFunction.java:58]:
```suggestion
                .features(Feature.DETERMINISTIC)
```

This isn't a strict nullable given that `select array_cat(null, array[1, 2]);` still returns a value. `nullable` can only be used if it returns null as soon as any argument is null. (We should probably rename it from `NULLABLE` to `STRICTNULL`)

Comment (seut) [reply server/src/main/java/io/crate/expression/scalar/ArrayCatFunction.java:58]: Hm but `array_cat(null, array[1, 2]);` would not match this function. What do I miss?

File: libs/sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java (hunk 2 above)
Comment (seut) [reply libs/sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java:2003]: Right 🤦 , good catch, thx.

File: server/src/main/java/io/crate/expression/scalar/ArrayCatFunction.java (hunk 4 above)
Comment (mfussenegger) [reply server/src/main/java/io/crate/expression/scalar/ArrayCatFunction.java:58]: Ah right - I was looking at the `evaluate` implementation, not at the `BinaryScalar` here.

File: server/src/main/java/io/crate/expression/scalar/ArrayCatFunction.java (hunk 3 above)
Comment (seut) [reply server/src/main/java/io/crate/expression/scalar/ArrayCatFunction.java:51]: > Would there be any downsides to allowing array(undefined) instead?

I see no real issue with allowing both arguments to be `array(undefined)` beside that it will deviate from PG's behaviour which raises and exception in this case. Other than this, it would just work and returns an empty array of type `array(undefined)`

Comment (mfussenegger) [reply server/src/main/java/io/crate/expression/scalar/ArrayCatFunction.java:51]: Ah I think what I missed was that we want to have the error for `array_cat([], [])` where the inner type is also undefined. 

Maybe rephrase slightly?

```
        // Registration to match `array_cat(null, null)` which should return `null`
        // Without this the signature below (`array_cat(array_undefined, array_undefined)`) would match
        // leading to a validation error which we need to preserve for `array_cat([], [])`
```

Comment (seut) [reply server/src/main/java/io/crate/expression/scalar/ArrayCatFunction.java]: > Ah I think what I missed was that we want to have the error for array_cat([], []) where the inner type is also undefined.

Do we really want the error?  I see no real need for it. I've changed it to not raise an error, see https://github.com/crate/crate/pull/16295/commits/6357e0b16b238061b76d11cf29b9896786c97de3

File: libs/sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java (hunk 2 above)
Comment (seut) [reply libs/sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java]: I did like you suggested, see https://github.com/crate/crate/pull/16295/commits/babf8e8a7990ec17a25ea742cfdf0f60fcc74bb1

File: server/src/main/java/io/crate/expression/scalar/ArrayCatFunction.java (hunk 3 above)
Comment (mfussenegger) [reply server/src/main/java/io/crate/expression/scalar/ArrayCatFunction.java]: Also okay I guess. I can't imagine a scenario where somebody would want to rely on that error.

File: server/src/main/java/io/crate/expression/scalar/ConcatFunction.java [hunk 5]
```diff
@@ -53,6 +54,17 @@ public static void register(Functions.Builder module) {
             StringConcatFunction::new
         );
 
+        // operator version of concat, it differs at it will return null if any of the arguments is null
+        module.add(
+            Signature.builder(OPERATOR_NAME, FunctionType.SCALAR)
+                .argumentTypes(DataTypes.STRING.getTypeSignature(),
+                    DataTypes.STRING.getTypeSignature())
+                .returnType(DataTypes.STRING.getTypeSignature())
+                .features(Feature.DETERMINISTIC, Feature.NON_NULLABLE)
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/scalar/ConcatFunction.java:63]:
```suggestion
                .features(Feature.DETERMINISTIC, Feature.NULLABLE)
```

**Review by mfussenegger — APPROVED** [review]
lgtm

Would be worth following up on adding a format special case in `Function.toString(Style style)`

File: server/src/main/java/io/crate/expression/scalar/ConcatFunction.java [hunk 6]
```diff
@@ -41,6 +41,7 @@
 public abstract class ConcatFunction extends Scalar<String, String> {
 
     public static final String NAME = "concat";
+    public static final String OPERATOR_NAME = "_concat_op";
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/scalar/ConcatFunction.java:44]: Could use `Operator.PREFIX + _concat` here for consistency? Although it shouldn't really matter if we keep it internal & hidden

Comment (mfussenegger) [reply server/src/main/java/io/crate/expression/scalar/ConcatFunction.java]: Looking at `Operator.PREFIX` usages and https://github.com/crate/crate/blob/15cfe452bba93bbcb2dfa10c6ac9e37627c78729/server/src/main/java/io/crate/expression/symbol/Function.java#L467-L481 I think going with `Operator.PREFIX + "||"` might make the formatting work out of the box but can do that in a follow up.

Comment (seut) [reply server/src/main/java/io/crate/expression/scalar/ConcatFunction.java]: As I forgot to register the operator name also for arrays and objects, I had to push a fixup anyhow. See https://github.com/crate/crate/pull/16295/commits/4d7f7ced3bebbb035a4bd71f5d678bb950cdad60

Comment (seut) [reply server/src/main/java/io/crate/expression/scalar/ConcatFunction.java]: I also added a formatting test. As you said, it works out-of-the-box when using the operator prefix.

File: server/src/main/java/io/crate/expression/scalar/ConcatFunction.java [hunk 7]
```diff
@@ -80,10 +82,43 @@ public static void register(Functions.Builder module) {
...
+                .typeVariableConstraints(typeVariable("E"))
+                .build(),
+            ArrayCatFunction::new
+        );
+        module.add(
+            Signature.builder(OPERATOR_NAME, FunctionType.SCALAR)
+                .argumentTypes(DataTypes.UNTYPED_OBJECT.getTypeSignature(),
+                    DataTypes.UNTYPED_OBJECT.getTypeSignature())
+                .returnType(DataTypes.UNTYPED_OBJECT.getTypeSignature())
+                .features(Feature.DETERMINISTIC, Feature.NON_NULLABLE)
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/scalar/ConcatFunction.java:117]:
```suggestion
                .features(Feature.DETERMINISTIC)
```

I think the nullability is conditional here - if I see it right this will return `null` if both arguments are null

Same for `array_cat` which now also returns null if all arguments are null

Comment (seut) [reply server/src/main/java/io/crate/expression/scalar/ConcatFunction.java:117]: Hm right, thought null will match the string version but this isn't true when the data type is set.

**Review by mfussenegger — APPROVED** [review]
